mqtt/mqtt-broker/src/broker.rs

```diff
@@ -42,33 +37,52 @@ where
 
 impl<N, Z> Broker<N, Z>
 where
-    N: Authenticator,
-    Z: Authorizer,
+    N: Authenticator + Send + 'static,
+    Z: Authorizer + Send + 'static,
 {
     pub fn handle(&self) -> BrokerHandle {
         BrokerHandle(self.sender.clone())
     }
 
-    pub async fn run(mut self) -> BrokerState {
-        while let Some(message) = self.messages.recv().await {
+    pub async fn run(self) -> Result<BrokerState, Error> {
+        let (tx, rx) = tokio::sync::oneshot::channel::<()>();
+        let handle = thread::Builder::new()
+            .name("mqtt::broker".to_string())
+            .spawn(|| {
+                let state = self.broker_loop();
+                if let Err(_e) = tx.send(()) {
+                    error!("failed to send broker state to event loop on shutdown");
+                }
+                state
+            })
+            .expect("failed to spawn broker thread");
+
+        // wait for the thread to exit
+        rx.await.map_err(Error::ThreadShutdown)?;
```

Comment:
Didn't quite understand, why we need oneshot channel to wait until the thread is finished, while we do the same with `handle.join()`?

Reply:
It's to propagate any panics that happen on the broker thread.
I'm pushing a change that handles panics better.

Reply:
The `oneshot` is used to signal the event loop that the thread has exited. The `handle.join()` is to propagate any panics from the broker thread onto the event loop thread (so that the process exits/crashes)